api/src/permissions/validateChangeReq.ts

```diff
@@ -0,0 +1,135 @@
+        const dbLangDoc = await dbService.getDocs([doc.language], [DocType.Language]);
+        if (dbLangDoc.docs.length > 0) {
+            const language = plainToInstance(LanguageDto, dbLangDoc.docs[0]);
+
+            // Get groups to which the user has Translate access to for Language documents
+            const userLanguageGroups = accessMap.calculateAccess(
+                [DocType.Language],
+                AclPermission.Translate,
+            );
+
+            for (const groupId of language.memberOf) {
+                if (!userLanguageGroups.includes(groupId)) {
+                    return "No 'Translate' access to the language of the Content object";
+                }
+            }
+        }
+
+        // Check if the user has access to set the publishStatus to Published
+        if (doc.status === PublishStatus.Published) {
+            const publishGroups = accessMap.calculateAccess(
+                [parentDoc.type],
+                AclPermission.Publish,
+            );
+            for (const groupId of parentDoc.memberOf) {
+                if (!publishGroups.includes(groupId)) {
+                    return "No 'Publish' access to document type 'Content'";
+                }
+            }
+        }
+    } else if (doc.memberOf && Array.isArray(doc.memberOf) && doc.memberOf.length > 0) {
+        // Check if user has edit access to any other types of documents
+        // -------------------------------------------------------------
+        const editGroups = accessMap.calculateAccess([doc.type], AclPermission.Edit);
+        for (const groupId of doc.memberOf) {
+            if (!editGroups.includes(groupId)) {
+                return "No 'Edit' access to one or more groups";
+            }
+        }
+    } else {
+        return "Unable to verify access. The document is not a group or does not have group membership";
+    }
+
+    // Validate tag assign access
+    // ============================
+    if (doc.tags) {
+        // Get tag documents from database
+        const tagDocs: MangoResponse<unknown> = await dbService.getDocs(doc.tags, [DocType.Tag]);
+
+        // Get array of groups to which the user has Assign access
+        const assignGroups = accessMap.calculateAccess([DocType.Tag], AclPermission.Assign);
+
+        // Compare tag group membership with groups to which the user has assign access to
+        if (tagDocs.docs && Array.isArray(tagDocs.docs)) {
+            for (const d of tagDocs.docs) {
+                const tagDoc = d as any;
+                if (tagDoc.memberOf && Array.isArray(tagDoc.memberOf)) {
+                    for (const groupId of tagDoc.memberOf) {
+                        if (!assignGroups.includes(groupId)) {
+                            return "No 'Assign' access to one or more tags";
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    return ""; // Empty response means the vaildation passed (i.e. no error)
```

Comment:
Consider returning a specific object with a boolean result and message to make it more obvious when validation succeeds or fails.

Reply:
Done